app/Platform/Services/GameListService.php

```diff
@@ -0,0 +1,519 @@
+                        return $reverse ? -1 : 1;
+                    }
+                } elseif ($b['achievements_published'] == 0) {
+                    return $reverse ? 1 : -1;
+                }
+
+                if ($a['NumTickets'] == $b['NumTickets']) {
+                    // same number of tickets; apply secondary sort on sort title
+                    return $reverse ? $b['SortTitle'] <=> $a['SortTitle'] : $a['SortTitle'] <=> $b['SortTitle'];
+                }
+
+                return $a['NumTickets'] <=> $b['NumTickets'];
+            },
+            'progress' => function ($a, $b) use ($reverse) {
+                if ($a['CompletionPercentage'] == $b['CompletionPercentage']) {
+                    // same completion progress; apply secondary sort on sort title, grouping sets with achievements first
+                    if ($a['achievements_published'] == 0) {
+                        if ($b['achievements_published'] != 0) {
+                            return $reverse ? -1 : 1;
+                        }
+                    } elseif ($b['achievements_published'] == 0) {
+                        return $reverse ? 1 : -1;
+                    }
+
+                    return $reverse ? $b['SortTitle'] <=> $a['SortTitle'] : $a['SortTitle'] <=> $b['SortTitle'];
+                }
+
+                return $a['CompletionPercentage'] <=> $b['CompletionPercentage'];
+            },
+        };
+
+        usort($this->games, $sortFunction);
+
+        if ($reverse) {
+            $this->games = array_reverse($this->games);
+        }
+    }
+
+    private function getTitleColumn(array $filterOptions): array
+    {
+        return [
+            'header' => 'Title',
+            'render' => function ($game) use ($filterOptions) {
+                if (!$filterOptions['console']) {
+                    echo '<td class="py-2">';
+                    echo Blade::render('
+                        <x-game.multiline-avatar
+                            :gameId="$ID"
+                            :gameTitle="$Title"
+                            :gameImageIcon="$ImageIcon"
+                            :consoleName="$ConsoleName"
+                        />', $game);
+                    echo '</td>';
+                } else {
+                    echo '<td>';
+                    echo Blade::render('
+                        <x-game.multiline-avatar
+                            :gameId="$ID"
+                            :gameTitle="$Title"
+                            :gameImageIcon="$ImageIcon"
+                        />', $game);
+                    echo '</td>';
+                }
+            },
+        ];
+    }
+
+    private function renderNumberOrBlank(array $game, string $field): void
+    {
+        if ($game[$field] == 0) {
+            echo '<td></td>';
+        } else {
+            echo '<td class="text-right">';
+            echo localized_number($game[$field]);
+            echo '</td>';
+        }
+    }
```

Comment:
These methods are highly functional, but also mix the presentational logic with the service logic.

The service should return data and let the Blade renderer handle presentation. For example:

```php
private function getTitleColumn(array $filterOptions): array
{
    return [
        'header' => 'Title',
        'render' => function ($game) use ($filterOptions) {
            return view('components.title-column', ['game' => $game, 'filterOptions' => $filterOptions])->render();
        }
    ];
}
```

IMO, each of these should request some Blade component to be rendered, and that component should handle its view concerns such as DOM template structure and what classnames to apply. Otherwise we are back to our legacy pattern of `echo`ing DOM.

Reply:
For cases where we're rendering just a single number or blank space, these existing methods are probably OK. It's likely just the `getBacklogColumn()` content that should be migrated to a Blade component. If the other columns become more complex at a later point in time, maybe that's the appropriate time for them to be migrated rather than now.

Reply:
I've moved the blacklog column into a blade component and left the simple text fields as echo statements.